warp10/src/main/java/io/warp10/script/functions/POLYFIT.java

```diff
@@ -0,0 +1,119 @@
+//   limitations under the License.
+//
+
+package io.warp10.script.functions;
+
+import io.warp10.continuum.gts.GTSHelper;
+import io.warp10.continuum.gts.GeoTimeSerie;
+import io.warp10.continuum.gts.GeoTimeSerie.TYPE;
+import io.warp10.script.NamedWarpScriptFunction;
+import io.warp10.script.WarpScriptException;
+import io.warp10.script.WarpScriptStack;
+import io.warp10.script.WarpScriptStackFunction;
+
+import java.util.Arrays;
+
+import org.apache.commons.math3.fitting.PolynomialCurveFitter;
+import org.apache.commons.math3.fitting.WeightedObservedPoints;
+
+/**
+ * Fit a polynomial curve to a GTS
+ */
+public class POLYFIT extends NamedWarpScriptFunction implements WarpScriptStackFunction {
+
+  public POLYFIT(String name) {
+    super(name);
+  }
+
+  @Override
+  public Object apply(WarpScriptStack stack) throws WarpScriptException {
+    Object top = stack.pop();
+
+    boolean useBucketId = false;
+
+    if (top instanceof Boolean) {
+      useBucketId = Boolean.TRUE.equals(top);
+      top = stack.pop();
+    }
+
+    if (!(top instanceof Long)) {
+      throw new WarpScriptException(getName() + " expects a polynomial degree on top if the stack.");
+    }
+
+    PolynomialCurveFitter fitter = PolynomialCurveFitter.create(((Number) top).intValue());
+
+    top = stack.pop();
+
+    if (!(top instanceof GeoTimeSerie)) {
+      throw new WarpScriptException(getName() + " expects a Geo Time Series below the degree of the polynomial.");
+    }
+
+    GeoTimeSerie gts = (GeoTimeSerie) top;
+
+    int n = GTSHelper.nvalues(gts);
+
+    if (0 == n) {
+      throw new WarpScriptException(getName() + " operates on a non empty Geo Time Series.");
+    }
+
+    if (useBucketId && !GTSHelper.isBucketized(gts)) {
+      throw new WarpScriptException(getName() + " use of bucket id is only possible for bucketized Geo Time Series.");
+    }
+
+    if (TYPE.LONG != gts.getType() && TYPE.DOUBLE != gts.getType()) {
+      throw new WarpScriptException(getName() + " can only operate on numerical Geo Time Series.");
+    }
+
+    //
+    // Build the list of observations
+    //
+
+    long firsttick = GTSHelper.firsttick(gts);
+    long bucketspan = GTSHelper.getBucketSpan(gts);
+
+    WeightedObservedPoints points = new WeightedObservedPoints();
+
+    for (int i = 0; i < n; i++) {
+      Object value = GTSHelper.valueAtIndex(gts, i);
+
+      long tick = GTSHelper.tickAtIndex(gts, i);
+
+      if (useBucketId) {
+        //
+        // Compute bucket id
+        //
+
+        int bucket = (int) ((tick - firsttick) / bucketspan);
+
+        points.add(bucket, ((Number) value).doubleValue());
+      } else {
+        points.add(tick, ((Number) value).doubleValue());
+      }
+    }
+
+    //
+    // Limit the number of iterations to twice the number of datapoints we attempt to fit
+    //
+
+    fitter.withMaxIterations(n * 2);
+
+    double[] coeffs = fitter.fit(points.toList());
+
+    stack.push(Arrays.asList(coeffs));
```

Comment:
This will leave an Arrays.ArrayList with a single double[] element in it, prefer:

```
    ArrayList<Double> coeffsList = new ArrayList<>(coeffs.length);
    for(double coeff: coeffs) {
      coeffsList.add(coeff);
    }

    stack.push(coeffsList);
```